Description: notifications: fix crash when removing monitor
Origin: upstream, https://gitlab.gnome.org/GNOME/gnome-flashback/commit/d91c04a2cf373f10
Last-Update: 2018-09-11

--- a/gnome-flashback/libnotifications/nd-stack.c
+++ b/gnome-flashback/libnotifications/nd-stack.c
@@ -436,10 +436,9 @@
         gtk_window_move (GTK_WINDOW (bubble), x, y);
 
         if (new_notification) {
-                g_signal_connect_swapped (G_OBJECT (bubble),
-                                          "destroy",
-                                          G_CALLBACK (nd_stack_remove_bubble),
-                                          stack);
+                g_signal_connect_object (bubble, "destroy",
+                                         G_CALLBACK (nd_stack_remove_bubble), stack,
+                                         G_CONNECT_SWAPPED);
                 stack->priv->bubbles = g_list_prepend (stack->priv->bubbles, bubble);
         }
 }
