diff --git a/gnome-flashback/libpolkit/flashback-polkit-dialog.c b/gnome-flashback/libpolkit/flashback-polkit-dialog.c
index aaeb5f3..9dc4de1 100644
--- a/gnome-flashback/libpolkit/flashback-polkit-dialog.c
+++ b/gnome-flashback/libpolkit/flashback-polkit-dialog.c
@@ -1048,6 +1048,8 @@ flashback_polkit_dialog_run_until_response_for_prompt (FlashbackPolkitDialog *di
   gtk_entry_set_text (GTK_ENTRY (dialog->password_entry), "");
   gtk_widget_grab_focus (dialog->password_entry);
 
+  g_object_set (G_OBJECT (dialog->password_entry), "im-module", "", NULL);
+
   if (was_cancelled != NULL)
     *was_cancelled = FALSE;
 
