diff --git a/gnome-flashback/libdesktop/gf-icon-view.c b/gnome-flashback/libdesktop/gf-icon-view.c
index b22602c..b7e46e7 100644
--- a/gnome-flashback/libdesktop/gf-icon-view.c
+++ b/gnome-flashback/libdesktop/gf-icon-view.c
@@ -1313,6 +1313,7 @@ create_popup_menu (GfIconView *self)
   GtkWidget *popup_menu;
   GtkStyleContext *context;
   GtkWidget *item;
+  GDesktopAppInfo *app_info;
 
   popup_menu = gtk_menu_new ();
 
@@ -1363,17 +1364,21 @@ create_popup_menu (GfIconView *self)
                     G_CALLBACK (display_settings_cb),
                     self);
 
-  item = gtk_separator_menu_item_new ();
-  gtk_menu_shell_append (GTK_MENU_SHELL (popup_menu), item);
-  gtk_widget_show (item);
-
-  item = gtk_menu_item_new_with_label (_("Open Terminal"));
-  gtk_menu_shell_append (GTK_MENU_SHELL (popup_menu), item);
-  gtk_widget_show (item);
-
-  g_signal_connect (item, "activate",
-                    G_CALLBACK (open_terminal_cb),
-                    self);
+  app_info = g_desktop_app_info_new ("org.gnome.Terminal.desktop");
+
+  if (app_info != NULL) {
+    item = gtk_separator_menu_item_new ();
+    gtk_menu_shell_append (GTK_MENU_SHELL (popup_menu), item);
+    gtk_widget_show (item);
+   
+    item = gtk_menu_item_new_with_label (_("Open Terminal"));
+    gtk_menu_shell_append (GTK_MENU_SHELL (popup_menu), item);
+    gtk_widget_show (item);
+   
+    g_signal_connect (item, "activate",
+                      G_CALLBACK (open_terminal_cb),
+                      self);
+  }
 
   return popup_menu;
 }
