diff --git a/gnome-flashback/libdesktop/gf-icon-view.c b/gnome-flashback/libdesktop/gf-icon-view.c
index e117f6b..0ebcc7a 100644
--- a/gnome-flashback/libdesktop/gf-icon-view.c
+++ b/gnome-flashback/libdesktop/gf-icon-view.c
@@ -3247,10 +3247,13 @@ gf_icon_view_validate_new_name (GfIconView  *self,
       name = gf_icon_get_name (GF_ICON (info->icon));
       if (g_strcmp0 (name, text) == 0)
         {
-          if (is_dir)
-            *message = g_strdup (_("A folder with that name already exists."));
-          else
-            *message = g_strdup (_("A file with that name already exists."));
+          if (g_strstr_len (text, -1, "/") == NULL)
+            {
+              if (is_dir)
+                *message = g_strdup (_("A folder with that name already exists."));
+              else
+                *message = g_strdup (_("A file with that name already exists."));
+            }
 
           valid = FALSE;
           break;
